src/System.Private.CoreLib/shared/System/Threading/WaitHandle.cs

```diff
@@ -0,0 +1,384 @@
+                    // not perfectly clear what the right behavior is here 
+                    // anyways.  This preserves Everett behavior.  We should 
+                    // ideally do these things:
+                    // *) Expose a settable SafeHandle property on WaitHandle.
+                    // *) Expose a settable OwnsHandle property on SafeHandle.
+                    if (_waitHandle != null)
+                    {
+                        _waitHandle.SetHandleAsInvalid();
+                        _waitHandle = null;
+                    }
+                }
+                else
+                {
+                    _waitHandle = new SafeWaitHandle(value, true);
+                }
+            }
+        }
+
+        public SafeWaitHandle SafeWaitHandle
+        {
+            get
+            {
+                if (_waitHandle == null)
+                {
+                    _waitHandle = new SafeWaitHandle(InvalidHandle, false);
+                }
+                return _waitHandle;
+            }
+
+            set
+            { _waitHandle = value; }
+        }
+
+        internal static int ToTimeoutMilliseconds(TimeSpan timeout)
+        {
+            var timeoutMilliseconds = (long)timeout.TotalMilliseconds;
+            if (timeoutMilliseconds < -1)
+            {
+                throw new ArgumentOutOfRangeException(nameof(timeout), SR.ArgumentOutOfRange_NeedNonNegOrNegative1);
+            }
+            if (timeoutMilliseconds > int.MaxValue)
+            {
+                throw new ArgumentOutOfRangeException(nameof(timeout), SR.ArgumentOutOfRange_LessEqualToIntegerMaxVal);
+            }
+            return (int)timeoutMilliseconds;
+        }
+
+        public virtual void Close() => Dispose();
+
+        protected virtual void Dispose(bool explicitDisposing)
+        {
+            if (_waitHandle != null)
+            {
+                _waitHandle.Close();
+            }
+        }
+
+        public void Dispose()
+        {
+            Dispose(true);
+            GC.SuppressFinalize(this);
+        }
+
+        public virtual bool WaitOne(int millisecondsTimeout)
+        {
+            if (millisecondsTimeout < -1)
+            {
+                throw new ArgumentOutOfRangeException(nameof(millisecondsTimeout), SR.ArgumentOutOfRange_NeedNonNegOrNegative1);
+            }
+
+            // The field value is modifiable via the public <see cref="WaitHandle.SafeWaitHandle"/> property, save it locally
+            // to ensure that one instance is used in all places in this method
+            SafeWaitHandle waitHandle = _waitHandle;
+            if (waitHandle == null)
+            {
+                // Throw ObjectDisposedException for backward compatibility even though it is not be representative of the issue
+                throw new ObjectDisposedException(null, SR.ObjectDisposed_Generic);
+            }
+
+            bool success = false;
+            waitHandle.DangerousAddRef(ref success);
```

Comment:
It is more correct to have `waitHandle.DangerousAddRef(ref success);` inside the `try`